### ISSUE TITLE
OpenGL sRGB state inconsistent starting with 2.29.1

### ISSUE THREAD

**Issue by anton-kl** [libsdl-org/SDL#8932]
Hello, it seems that on my setup starting with SDL 2.29.1 when I request OpenGL context, for several rgba+srgb flag setup I don't get the srgb I requested via `SDL_GL_FRAMEBUFFER_SRGB_CAPABLE`, even though on 2.28.5 it works fine after repeated testing. For example I request RBGA 8,8,8,0 with sRGB capability off, but I receive context with sRGB on.

I've tested four versions with four test cases (each a couple times):

- `release-2.28.5` - OK
- `prerelease-2.29.1` - FAIL
- `prerelease-2.29.2` - FAIL
- `release-2.30.x` - FAIL

For me, starting with 2.29.1 I consistently receive wrong result for rgba=8,8,8,0 with srgb off - srgb is for some reason on, and for rgba=10,10,10,2 with srgb on - for some reason srgb is off.

I'm only reporting this since it's working fine on 2.28.5 - so I suspect regression, it doesn't seem like the intended behavior.

My test setup:
- NVIDIA GeForce RTX 3080 Ti (driver: 31.0.15.4633 (NVIDIA 546.33) DCH / Win11 64)
- Windows 11
- AMD Ryzen 7 3700X

Test code:
```c++
#include <SDL.h>
#include <GL/glew.h>
#include <iostream>
#include <format>

#define SDL_CHECK( x ) if (x != 0) { printf("Error %s:%d : %s\n", __FILE__, __LINE__, SDL_GetError()); return 1; } 

int main(int argc, char **argv) {
    if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_GAMECONTROLLER) != 0) {
        printf("Error: %s\n", SDL_GetError());
        return 1;
    }

    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, 0) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 4) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 3) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, 24) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_STENCIL_SIZE, 8) );

    int red_in = 8;
    int green_in = 8;
    int blue_in = 8;
    int alpha_in = 0;
    int srgb_in = 0;

    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_RED_SIZE, red_in) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_GREEN_SIZE, green_in) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_BLUE_SIZE, blue_in) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_ALPHA_SIZE, alpha_in) );
    SDL_CHECK( SDL_GL_SetAttribute(SDL_GL_FRAMEBUFFER_SRGB_CAPABLE, srgb_in) );

    auto m_sdlWindow = SDL_CreateWindow(
        "Hi",
        SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED,
        1920, 1080,
        SDL_WINDOW_OPENGL | SDL_WINDOW_ALLOW_HIGHDPI | SDL_WINDOW_RESIZABLE);
    if (!m_sdlWindow) { printf("Failed to create the window!: %s\n", SDL_GetError()); return 1; }

    auto m_glContext = SDL_GL_CreateContext(m_sdlWindow);
    SDL_CHECK( SDL_GL_MakeCurrent(m_sdlWindow, m_glContext) );

    int red_out = 0;    SDL_CHECK( SDL_GL_GetAttribute(SDL_GL_RED_SIZE, &red_out) );
    int green_out = 0;  SDL_CHECK( SDL_GL_GetAttribute(SDL_GL_GREEN_SIZE, &green_out) );
    int blue_out = 0;   SDL_CHECK( SDL_GL_GetAttribute(SDL_GL_BLUE_SIZE, &blue_out) );
    int alpha_out = 0;  SDL_CHECK( SDL_GL_GetAttribute(SDL_GL_ALPHA_SIZE, &alpha_out) );
    int srgb_out = 0;   SDL_CHECK( SDL_GL_GetAttribute(SDL_GL_FRAMEBUFFER_SRGB_CAPABLE, &srgb_out) );

    SDL_version verCompiled;
    SDL_VERSION(&verCompiled);
    SDL_version verLinked;
    SDL_GetVersion(&verLinked);
    std::cout << std::format("SDL compiled with: {}.{}.{}, linked against: {}.{}.{}\n",
        verCompiled.major, verCompiled.minor, verCompiled.patch,
        verLinked.major, verLinked.minor, verLinked.patch);
    std::cout << std::format("Requested: {},{},{},{} srgb: {}\n",
        red_in, green_in, blue_in, alpha_in, srgb_in);
    std::cout << std::format("Received: {},{},{},{} srgb: {}\n",
        red_out, green_out, blue_out, alpha_out, srgb_out);

    /*
    For SDL 2.28.5 everything is consistently the same (requested = received,
      except for the last case where we get different bit size for RGBA but this is fine):
    SDL compiled with: 2.28.5, linked against: 2.28.5
        - Requested: 8,8,8,0 srgb: 0
          Received: 8,8,8,0 srgb: 0
        - Requested: 8,8,8,0 srgb: 1
          Received: 8,8,8,0 srgb: 1
        - Requested: 10,10,10,2 srgb: 0
          Received: 10,10,10,2 srgb: 0
        - Requested: 10,10,10,2 srgb: 1
          Received: 16,16,16,16 srgb: 1

    For 2.29.1+ we start seeing SRGB unexpectedly turn on for one configuration, being off for another
      (all three tested versions behave the same):
    SDL compiled with: 2.29.1, linked against: 2.29.1
    SDL compiled with: 2.29.2, linked against: 2.29.2
    (release-2.30.x 76e64053) SDL compiled with: 2.29.2, linked against: 2.29.2
        - Requested: 8,8,8,0 srgb: 0
          Received: 8,8,8,0 srgb: 1 // sRGB FAIL
        - Requested: 8,8,8,0 srgb: 1
          Received: 8,8,8,0 srgb: 1
        - Requested: 10,10,10,2 srgb: 0
          Received: 10,10,10,2 srgb: 0
        - Requested: 10,10,10,2 srgb: 1
          Received: 16,16,16,16 srgb: 0 // sRGB FAIL
    */

    return 0;
}
```

**Comment by slouken** [comment]
Yes, this sounds like a regression. Are you able to git bisect to a specific commit?

**Comment by slouken** [comment]
I suspect it might be 594a79c2f9db21a11a76c012b2b1655a21fc8982, where we're now reporting whether you got an sRGB visual. It may be that you weren't before and didn't know it?

**Comment by anton-kl** [comment]
Oh no, I'm for sure getting the wrong sRGB - can confirm visually.
I've reverted https://github.com/libsdl-org/SDL/commit/594a79c2f9db21a11a76c012b2b1655a21fc8982 and it fixed the problem! So there must be some issue in the code

**Comment by anton-kl** [comment]
I've checked in render doc to be 100% sure that the values reported are what I see, and I can see that they are - I get `0 Backbuffer Color Texture 2D 2462 1401 1 1 R8G8B8A8_SRGB (GL_FRAMEBUFFER_SRGB = 0)` with https://github.com/libsdl-org/SDL/commit/594a79c2f9db21a11a76c012b2b1655a21fc8982 included, and `0 Backbuffer Color Texture 2D 2462 1401 1 1 R8G8B8A8_SRGB` with it reverted. So for me it seem like reporting was correct before the commit and stays correct after - but the behavior changes.

I have to go, but I can dive deeper into the code tomorrow

**Comment by anton-kl** [comment]
Alright, after some investigation I came up to the conclusion that I'm dealing with a driver bug, and there is no issue on the SDL side per se. For some input driver just reports wrong data, and before the fix, by ignoring driver's report, it seemed like SDL was reporting correctly. Now after the fix SDL is always taking what driver says, even if it's wrong, which is why it seemed that there is an issue. Instead, it just randomly unveiled a driver-side problem (at least as far as I can tell).

For example, for request `10,10,10,2 srgb: 1` the driver tells us that we received an sRGB-incapable framebuffer `16,16,16,16 srgb: 0` (which is also reported by SDL after the fix), but instead the opposite is true: this particular pixel format for some reason "locked" to sRGB on my setup, in a sense that I can't disable it via `glDisable(GL_FRAMEBUFFER_SRGB)`, and I have no way of knowing it is the situation - `wglGetPixelFormatAttribivARB` never returns `WGL_FRAMEBUFFER_SRGB_CAPABLE_ARB=true` for whatever pixel format we are selecting, GL_FRAMEBUFFER_SRGB is not set. So it appears like sRGB is disabled and not supported, instead the driver does the sRGB conversion (which has to be mitigated).

Before the fix SDL was reporting sRGB=1 in this case (SDL would just keep the value we requested and won't update it), which was closer to the truth of what was happening, but incorrect when it comes to what the driver reports (which is sRGB=0, even if it's wrong). The situation is nasty because after the fix you think you shouldn't do any sRGB correction (and can't), while in reality you must - the image will look wrong otherwise. 

Same situation with `11,11,10,2 srgb: 0` -- we are being locked in into sRGB (enabling/disabling GL_FRAMEBUFFER_SRGB doesn't change anything), driver repots no sRGB capability instead. The only difference that in this case we didn't even request sRGB.

I found some topics online that mention somewhat similar problems: https://forums.developer.nvidia.com/t/gl-framebuffer-srgb-functions-incorrectly/34889

@slouken
To sum it up, I think this is no issue and sorry if I wasted some of your time :)
I don't think there will be many people upgrading to a newer SDL version, using OpenGL, requesting sRGB buffer, using unusual (>8) bit-sizes for the default framebuffer, and being surprised by new report all at the same time. And it seems there is no plausible workaround - either keep the old behavior and introduce new one via a separate API (bad), or do more hackery/hardcoding for specific gpus/drivers/cases (not realistic).

**Comment by slouken** [comment]
I'm sorry this didn't work out, but that was great investigation, thanks!
